### ISSUE TITLE
Windows: Slow resolution due to querying of non-existing autoconf IPv6 servers

### ISSUE THREAD

**Issue by nmaier** [c-ares/c-ares#134]
Affected
---
- c-ares 0.13.0 w/ system DNS server list
- Windows 7 (other Windows versions not tested, but potentially affected as well)

Problem
---
c-ares regressed in lookup time from a few milliseconds to >10s.

Pesky IPv6 DNS Servers (or: The cause)
---
Starting with #81 (or rather 1.13.0) I reliably and always observe massive delays from c-ares (via https://github.com/aria2/aria2/) running on Windows 7 when letting c-ares pick the name servers itself by querying the Windows configuration.

I tried looking into it, and c-ares now picks up 3 additional IPv6 DNS server addresses:
- fec0:0:0:ffff::1
- fec0:0:0:ffff::2
- fec0:0:0:ffff::3

Those are site-local addresses that get autoconfigured in Windows for IPv6 whenever a NIC has no DHCP or static configured DNS servers, but IPv6 connectivity in theory. It seems it's Microsoft's attempt trying to implement https://tools.ietf.org/html/draft-ietf-ipv6-dns-discovery-07 (which they also authored)

So where are these coming from exactly?
Adding some `printf`s to `get_DNS_AdaptersAddresses` shows those DNS servers are configured for the "Loopback Pseudo-Interface 1" and "VirtualBox Host-Only Network" interfaces and picked up by c-ares from there.

It gets worse... c-ares "detects" a metric of `-1` for all DNS servers incl. the non-functioning ones, and `qsort`reliably sorts the bad ones first in my configuration, causing c-ares to first query three servers on sitelocal hosts that do not even exist, resulting in **multi-second delays of up to 14 seconds** in my tests. The final dns server string c-ares generates for me is:
`fec0:0:0:ffff::1,fec0:0:0:ffff::2,fec0:0:0:ffff::3,127.0.0.1`

Remedies
---

- Do not even consider the loopback interface, e.g. by amending the check for not-up interfaces to also skip loopbacks.
```c
    if (ipaaEntry->OperStatus != IfOperStatusUp ||
        ipaaEntry->IfType == IF_TYPE_SOFTWARE_LOOPBACK)
        continue;
```
- Addressing virtual interfaces such as "VirtualBox Host-Only Network" is a bit harder, tho
  - Ignore all interfaces without a `GatewayAddress`?
  - Explicitly blacklist those three IPv6 addresses?
- Or revert to the old behavior of "`GetNetworkParams` first"...

cc author and committer of #81 @bradh352 @bagder

**Comment by bradh352** [comment]
I'll see if I can submit a fix for this, this week.  Its not too difficult, just need to add an initial order index to the Address array to make qsort stable.  That way it gets the benefit of the old new logic, as well as not regressing from the old behavior.

We could entertain omitting loopback interfaces, but the old method didn't so I'm a little hesitant to do that without fully thinking through other implications.

**Comment by nmaier** [comment]
Actually no, stable qsort will just order DNS servers by whatever order Windows will return the interface information, so the bogus site-local autoconf servers might be still front.
Also, it does not help when a library consumer specifies `ARES_OPT_ROTATE`.

**Comment by nmaier** [comment]
Adding the bogus servers is a regression from the old behavior in itself btw. The `GetNetworkParams` path does not add the `fec0::` servers in the first place.

**Comment by bradh352** [comment]
GetNetworkParams didn't ever add ipv6 servers period, not even legitimate ones.  That is not acceptable behavior.

**Comment by nmaier** [comment]
Be that as it may, adding bogus ipv6 servers is not acceptable behavior either ;)

**Comment by gvanem** [comment]
@nmaier ... and qsort reliably sorts the bad ones first in my configuration

I too also have a similar setup as you. But the metrics should be compared as `ULONG` so `qsort()` and `compareAddresses()` should put those *dead servers*  last. How come it doesn't for you?

But I've seen `qsort()` behaving unreliably when all metrics are equal. Anyway, here is the string I get from `get_DNS_AdaptersAddresses()`:
```
8.8.8.8,193.213.112.4,fec0:0:0:ffff::1,fec0:0:0:ffff::2,fec0:0:0:ffff::3
```
@nmaier Do not even consider the loopback interface

That would be a bad idea for people who wants to use a local DNS-server.

**Comment by nmaier** [comment]
@gvanem because all my servers have a metric of -1 according to c-ares.
This is because all servers are kinda local I think (or does it always return -1 in Win7... didn't check). The bogus ones are site-local, my actual server is actually 127.0.0.1 on that system, and 127.0.0.1 gets reliably sorted last on my system, because while `qsort` is not stable, it is still deterministic and so is the Windows API to query interfaces on consecutive uses.

As such, I want to clarify my point about the loopback interface: I do not mean to exclude the loopback interface (or rather 127.0.0.0/8, ::1 or the hostnames for those which are usually "localhost") as a DNS possible server, but not query the actual virtual software loopback NIC (`Loopback Pseudo-Interface 1`) for DNS servers.
The DNS servers the actual virtual loopback NIC lists cannot be changed (that interface is not even listed in most user-facing tools, it isn't in the Network Center/Adapters nor `ipconfig` for example) and are always no IPv4 and the three bogus IPv6 servers configured for it, at least on Windows 7.
If you, like me actually, want to use a a local DNS Server, e.g. listening on 127.0.0.1, you'd not set it on that virtual loopback NIC but one of your other real or pseudo interfaces (e.g. I got it configured on my physical Gigabit NIC)

**Comment by bradh352** [comment]
Anyone that is experience this issue that wants to test b4053e4 would be appreciated.  I haven't been able to reproduce.

**Comment by nmaier** [comment]
@bradh352  it does really not help. Now the order is dependent on the order of interfaces returned by the Windows API. What's more, even if you get lucky* and the interfaces with non-bogus DNS servers are returned first, stuff still breaks in horrible ways when the software is using `ARES_OPT_ROTATE`.
Your whole patch not only does not address the issue of those bogus servers, furthermore all it really does it preserve the "sort" order of the original Windows API, which is nothing more than an implementation detail and hence cannot be relied upon not to change in breaking ways in the future (actually, I am not even sure the current Windows API implementation is really deterministic).

* I did get lucky, my generated string is now `127.0.0.1,fec0:0:0:ffff::1,fec0:0:0:ffff::2,fec0:0:0:ffff::3`, tho it still could have easily been otherwise if the Windows API decided to return interfaces in a different order.

**Comment by bradh352** [comment]
There was a comment left on the git commit itself that essentially said that preserving the original order would fix the issue for them, which is why I went this route.  So it should at least work around some of the issues people are having, but that was related to ipv4:
https://github.com/c-ares/c-ares/commit/39aeafd27ddabfcdcaa406ffb585d918bee7263e#commitcomment-22682305

That said, there's no way we can actually key off of the interface type to exclude it since that would have undesirable side-effects, so we'd need *something* else to sort by, or some way to know to exclude an address.  Of course we could *hack* it and exclude fec0::/10 prefixes (and possibly others) completely for this particular issue.

I did notice that powershell has Get-NetPrefixPolicy which lists a "precedence" for ipv6 prefixes (higher is more preferred), if we could get this list via C, we could choose the most specific match and use the precedence value for sorting.

**Comment by bradh352** [comment]
looks like it was @gvanem that make the git commit comment on this thread :)  So at least that issue should be resolved.

**Comment by nmaier** [comment]
Having thought about it a bit, I think the best route would be to just blacklist `fec0::` unless those are the only servers available. That way the number of potentially aversely affected users would be at a minimum:
- Users who use this type of autoconf on Windows (should be incredibly low to begin with as it's a MS thing that never got beyond RFC draft stage) or configured their DNS servers to use `fec0::` addresses, which technically is a reserved special range and shouldn't be used for manual stuff in the first place
- AND who furthermore also use static/DHCP DNS servers on other NICs (which makes it even a lot lower then).

Compared to now, where every Windows user is potentially affected (depending on the NIC order and `ARES_OPT_ROTATE`), this is a good tradeoff in my book

**Comment by bradh352** [comment]
I think it would be safe to exclude fec0::/10 and possibly some others.  I'll look into seeing how clean it would be to do this.

**Comment by bradh352** [comment]
Fixed in #144
